flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java

```diff
@@ -260,50 +293,66 @@ public String getMetric(String name, String defaultValue) {
 				? value
 				: defaultValue;
 		}
-	}
 
-	/**
-	 * Sub-structure containing metrics of the JobManager.
-	 */
-	public static class JobManagerMetricStore extends ComponentMetricStore {
+		public static ComponentMetricStore unmodifiable(ComponentMetricStore source) {
+			if (source == null) {
+				return null;
+			}
+			return new ComponentMetricStore(unmodifiableMap(source.metrics));
+		}
 	}
 
 	/**
 	 * Sub-structure containing metrics of a single TaskManager.
 	 */
+	@ThreadSafe
 	public static class TaskManagerMetricStore extends ComponentMetricStore {
-		public final Set<String> garbageCollectorNames = new HashSet<>();
+		public final Set<String> garbageCollectorNames;
+
+		public TaskManagerMetricStore() {
+			this(new ConcurrentHashMap<>(), ConcurrentHashMap.newKeySet());
+		}
+
+		public TaskManagerMetricStore(Map<String, String> metrics, Set<String> garbageCollectorNames) {
+			super(metrics);
+			this.garbageCollectorNames = checkNotNull(garbageCollectorNames);
+		}
 
 		public void addGarbageCollectorName(String name) {
 			garbageCollectorNames.add(name);
 		}
+
+		public static TaskManagerMetricStore unmodifiable(TaskManagerMetricStore source) {
```

Comment:
I'm wondering whether we really need this unmodifiable business. Yes, it's technically a good idea, but the access to the MetricStore is limited and fully under our control; so we _know_ that we never try to modify the map.

Reply:
It's a matter of clean design and maintaining it in the future. You only *know* that we do not modify this map **at the moment**. Are you sure that every committer reviewing changes in this parts of the code will always spot any such modifications without `unmodifiable` assertion?  On the other hand removing `unmodifiable` in a PR is big yellow warning light that's something is going on.

Following this logic, why don't we make all methods/fields `public`, since we know that we will never try to access supposed to be `private` fields?

Reply:
In which case we should rather modify the store to not allow writes in the first place, instead of opting for unmodifiable collections that are pretty much a hack. "here, have an object that fails for 50% of the defined methods"; that's hardly good design is it.

Till suggested dedicated read methods that return metrics, something like `List<String> getMetrics(List<String> metricNames)` instead of exposing the metric maps.

This would make the interface cleaner and would allow us to simplify the synchronization.

Reply:
Would you rather have a runtime error about illegal/unintended write access or none at all? Java sucks in this regard (missing const-correctness) :(

Exposing `List<...>` have the same issue.

Reply:
If the returned list is a fresh copy independent of the map it can be modifiable; the user can do whatever he wants with it.

Reply:
Then you pay costs of copying all of the objects, which is much higher compared to the synchronisation. But yes, you can do that. In that case you don't need concurrent hash maps. 

If you intend to go in other direction with this and leave the current code in the master as it is, please log the work in Jira for this after the release (I will close release blocker issue for which I started this PR). However I have an impression that any follow up work would be easier on top of this change and in the mean time this is still way better then current code on master branch with this external synchronisation.